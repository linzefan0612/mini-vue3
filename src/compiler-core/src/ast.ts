/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-10 10:45:42
 * @LastEditTime: 2022-04-10 12:29:06
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\ast.ts
 *
 */

export const enum NodeType {
  INTERPOLATION,
  SIMPLE_EXPRESSION,
  ELEMENT,
  TEXT,
}
