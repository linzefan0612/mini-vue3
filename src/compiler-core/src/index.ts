/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-20 20:49:37
 * @LastEditTime: 2022-04-20 20:49:37
 * @LastEditors: Lin ZeFan
 * @Description: 
 * @FilePath: \mini-vue3\src\compiler-core\src\index.ts
 * 
 */

// 暴露整个 compiler
export * from './compiler'