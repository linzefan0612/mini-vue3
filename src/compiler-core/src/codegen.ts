/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-17 10:40:19
 * @LastEditTime: 2022-04-17 16:32:08
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\codegen.ts
 *
 */

import { NodeType } from "./ast";
import {
  CREATE_ELEMENT_VNODE,
  HelperNameMapping,
  TO_DISPLAY_STRING,
} from "./runtimeHelpers";
import { isArray, isString } from "./utils";

export function codegen(ast) {
  const context = createCodeGenContext();
  const { push } = context;

  // 处理头部函数引入
  if (ast.helpers.length) {
    genFunctionPreamble(ast, context);
  }

  const funcName = "render";
  push(`export `);
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");
  push(`function ${funcName}(${signature}) { `);

  push(`return `);
  genNode(ast.codegenNode, context);
  push(` }`);
  return context.code;
}

function genNode(node, context) {
  // 根据 node 的类型进行不同的处理
  switch (node.type) {
    case NodeType.TEXT:
      genText(node, context);
      break;
    case NodeType.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeType.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeType.ELEMENT:
      genElement(node, context);
      break;
    case NodeType.COMPOUND_EXPRESSION:
      genCompoundExpression(node, context);
  }
}

function genExpression(node, context) {
  // 处理 SIMPLE_EXPRESSION
  const { push } = context;
  push(`'${node.content}'`);
}

function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  // 前面处理插值类型的时候，真正的值是包在content.content里的
  // { type: NodeType.INTERPOLATION, content: { type: NodeType.SIMPLE_EXPRESSION, content: 'message'} }
  genNode(node.content, context);
  push(`)`);
}

function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function genElement(node, context) {
  const { push, helper } = context;
  const { tag, children, props } = node;
  // push(`${helper(CREATE_ELEMENT_VNODE)}('${tag}')`);
  push(`${helper(CREATE_ELEMENT_VNODE)}(`);
  // 批量处理 tag，props 和 children，优化空值情况
  genNodeList(genNullable([tag, props, children]), context);
}

function genNodeList(nodes, context) {
  const { push } = context;
  for (let i = 0; i < nodes.length; i++) {
    const node = nodes[i];
    /** 处理node list
     * 1. 如果是text，直接拼接
     * 2. 如果是数组，遍历数组，把每一项再通过 genNode 检测类型
     * 3. 如果是对象，给 genNode 检测类型
     */
    if (isString(node)) {
      push(node);
    } else if (isArray(node)) {
      for (let j = 0; j < node.length; j++) {
        const n = node[j];
        genNode(n, context);
      }
    } else {
      genNode(node, context);
    }
    // 遍历完，加上分隔符
    i < nodes.length - 1 && push(", ");
  }
}

function genNullable(args) {
  // 把undefined、null，转为 “null”
  return args.map((arg) => arg || "null");
}

function genCompoundExpression(node, context) {
  const { children } = node;
  const { push } = context;
  // 对 children 进行遍历
  for (let i = 0; i < children.length; i++) {
    const child = children[i];
    // 如果是 string，也就是我们手动添加的 +
    if (isString(child)) {
      // 直接 push
      push(child);
    } else {
      // 否则还是走 genNode
      genNode(child, context);
    }
  }
}

function createCodeGenContext() {
  const context = {
    code: "",
    helper(key) {
      return `_${HelperNameMapping[key]}`;
    },
    push(source: string) {
      context.code += source;
    },
    addLine() {
      context.code += "\n";
    },
  };
  return context;
}
function genFunctionPreamble(ast: any, context) {
  // 引入都来自 Vue
  const VueBinding = "Vue";
  const { push, addLine } = context;
  // 因为是Symbol，需要用映射表匹配
  const aliasHelper = (s) =>
    `${HelperNameMapping[s]} as _${HelperNameMapping[s]}`;
  // 处理头部引入
  push(`const { ${ast.helpers.map(aliasHelper).join(", ")} } = ${VueBinding}`);
  addLine();
}
