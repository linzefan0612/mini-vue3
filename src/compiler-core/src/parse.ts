/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-10 10:45:42
 * @LastEditTime: 2022-04-16 16:56:00
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\parse.ts
 *
 */

import { NodeType } from "./ast";

// const DOM_TAG_REG = /^<\/?([a-z]*>$)/;
// const ELEMENT_REG =  /^<[a-z]*>$/

export function baseParse(content: string) {
  const context = createContext(content);
  return createRoot(parseChildren(context, []));
}

// 创建上下文
function createContext(content: string) {
  return {
    source: content,
  };
}

// 创建 ast 根节点
function createRoot(children) {
  return {
    children,
  };
}

// 创建 children
function parseChildren(context: { source: string }, ancestors): any {
  const nodes: any = [];
  while (isEnd(context, ancestors)) {
    let node;
    const s = context.source;
    /** 判断字符串类型
     * 1. 为插值
     * 2. 为element
     */
    if (s.startsWith("{{")) {
      // {{ 开头，即认为是插值
      node = parseInterpolation(context);
    } else if (s.startsWith("<") && /[a-z]/i.test(s[1])) {
      // <开头，并且第二位是a-z，即认为是element类型
      node = parseElement(context, ancestors);
    } else {
      node = parseText(context);
    }
    nodes.push(node);
  }
  return nodes;
}

function parseTextData(context: { source: string }, length) {
  return context.source.slice(0, length);
}
function advanceBy(context, length: number) {
  context.source = context.source.slice(length);
}

// 解析插值表达式
function parseInterpolation(context: { source: string }) {
  // 插值开始字符
  const openDelimiter = "{{";
  // 插值结束字符
  const closeDelimiter = "}}";
  // 找到插值结束字符的位置
  const closeIndex = context.source.indexOf(
    closeDelimiter,
    openDelimiter.length
  );
  // 切割掉 {{
  advanceBy(context, openDelimiter.length);
  // 找到 }} 前的内容
  const rawContentLength = closeIndex - closeDelimiter.length;
  // 截取插值表达式里的内容
  const content = parseTextData(context, rawContentLength).trim();
  // 完成{{}}的内容匹配，切割掉匹配完成的内容，继续往前推进，解析后面的内容
  advanceBy(context, rawContentLength + closeDelimiter.length);
  return {
    type: NodeType.INTERPOLATION,
    content: {
      type: NodeType.SIMPLE_EXPRESSION,
      content,
    },
  };
}

const enum TagType {
  START,
  END,
}

// 解析element
function parseElement(context: { source: string }, ancestors): any {
  // 这里调用两次 parseTag 处理前后标签
  const element: any = parseTag(context, TagType.START);
  // 收集标签
  ancestors.push(element);
  // 增加 parseChildren，储存包裹的内容
  element.children = parseChildren(context, ancestors);
  // 循环结束，把当前tag弹出标签，进入下一个循环
  ancestors.pop();
  /** 切除闭合标签
   * 1. 当前source标签内容等于首部tag，说明是闭合标签，则进行切除
   * 2. 不相等，则说明没有写闭合标签，报警告
   */
  if (startsWithEndTagOpen(context.source, element.tag)) {
    // 处理闭合标签
    parseTag(context, TagType.END);
  } else {
    throw new Error(`不存在结束标签：${element.tag}`);
  }
  return element;
}

function parseTag(context: { source: string }, type: TagType) {
  // i 忽略大小写, ([a-z]*) 作为一个分组，匹配<开头或者</开头的内容
  const match = /^<\/?([a-z]*)/i.exec(context.source);
  // 其中 tag[1] 就是匹配出来的 div，取反是为了避免为null报错
  const tag = match![1];
  /** 往后推进
   * 1. match[0]匹配出 <div
   * 2. match[0].length + 1，即下标<div>后开始
   */
  advanceBy(context, match![0].length + 1);
  // 处理闭合标签，就不需要return了
  if (type === TagType.END) return;
  return {
    tag,
    type: NodeType.ELEMENT,
  };
}

// 解析text
function parseText(context: { source: string }): any {
  const s = context.source;
  let len = s.length;
  /** 处理element包裹情况
   * 1. 新建一个TAG_ARRAY，用来判断text后可能存在的符号
   * 2. 取最贴近text的符号，因为 < 跟 {{ 可能同时都存在，取最小的，即离text内容最近的
   */
  const TAG_ARRAY = ["<", "{{"];
  for (let i = 0; i < TAG_ARRAY.length; i++) {
    const tag = TAG_ARRAY[i];
    const index = s.indexOf(tag);
    /** 获取text的位置
     * 1. 如果符号存在，并且小于len，取离text最近的内容
     * 例如 hi,</p>{{message}}，会先找到 < 的位置，覆盖len，又找到 {{，但是 {{ 比 len大，说明 {{ 符号在后面，所以不赋值
     * 2. 如果不存在，直接切到最后面即可
     */
    if (index !== -1 && index < len) {
      len = index;
    }
  }
  // 获取当前字符串内容
  const content = parseTextData(context, len);
  // 推进
  advanceBy(context, len);
  return {
    type: NodeType.TEXT,
    content,
  };
}

// 匹配是否结束标签
function isEnd(context: { source: string }, ancestors: any) {
  const s = context.source;
  /** 判断tag是结束标签
   * 1. 判断是否标签，是标签进入循环
   * 2. 从栈顶开始循环，栈是先入后出的，所以根标签会在最底部
   * 3. 判断当前的标签的tag是否跟栈的tag相等，相等则说明当前tag内容已经推导结束，结束当前children循环，进入下一个循环
   */
  if (s.startsWith("</")) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      // 如果说栈里存在这个标签，那么就跳出循环
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return false;
      }
    }
  }
  // 返回内容本身
  return s;
}

function startsWithEndTagOpen(source, tag) {
  const endTokenLength = "</".length;
  return source.slice(endTokenLength, tag.length + endTokenLength) === tag;
}
