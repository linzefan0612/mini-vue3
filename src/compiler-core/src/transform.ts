/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-16 22:33:45
 * @LastEditTime: 2022-04-17 16:18:02
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\transform.ts
 *
 */

import { NodeType } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createdRootCodegen(root);
  // 储存所有头部引入
  root.helpers = [...context.helpers.keys()];
}

function createTransformContext(root: any, options: any) {
  const { nodeTransforms = [] } = options;
  const context = {
    root,
    nodeTransforms: nodeTransforms || [],
    helpers: new Map(),
    helper(name: string) {
      this.helpers.set(name, 1);
    },
  };
  return context;
}

function traverseNode(node, context) {
  const { nodeTransforms } = context;
  const exitFns: any = [];
  for (let index = 0; index < nodeTransforms.length; index++) {
    const transform = nodeTransforms[index];
    const exitFn = transform(node, context);
    // 收集退出函数
    if (exitFn) exitFns.push(exitFn);
  }

  // 在这里遍历整棵树的时候，将根据不同的 node 的类型存入不同的 helper
  switch (node.type) {
    case NodeType.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;
    case NodeType.ROOT:
    case NodeType.ELEMENT:
      // 只有在 ROOT 和 ELEMENT 才会存在 children
      traverseChildren(node, context);
      break;
    default:
      break;
  }

  let i = exitFns.length;
  // 执行所有的退出函数
  while (i--) {
    exitFns[i]();
  }
}

function traverseChildren(node: any, context: any) {
  const { children } = node;
  for (let i = 0; i < children.length; i++) {
    traverseNode(children[i], context);
  }
}
function createdRootCodegen(root: any) {
  const child = root.children[0];
  // 在这里进行判断，如果说 children[0] 的类型是 ELEMENT，那么直接修改为 child.codegenNode
  if (child.type === NodeType.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}
