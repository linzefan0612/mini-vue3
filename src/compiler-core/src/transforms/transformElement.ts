/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-10 10:45:42
 * @LastEditTime: 2022-04-17 16:07:26
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\transforms\transformElement.ts
 *
 */
import { NodeType } from "../ast";
import { CREATE_ELEMENT_VNODE } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type === NodeType.ELEMENT) {
    // 添加相关的helper
    context.helper(CREATE_ELEMENT_VNODE);

    // 中间处理层，处理 props 和 tag
    const vnodeTag = node.tag;
    const vnodeProps = node.props;

    const { children } = node;
    let vnodeChildren = children;

    const vnodeElement = {
      type: NodeType.ELEMENT,
      tag: vnodeTag,
      props: vnodeProps,
      children: vnodeChildren,
    };

    node.codegenNode = vnodeElement;
  }
}
