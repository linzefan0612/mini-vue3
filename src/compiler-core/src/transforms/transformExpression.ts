/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-10 10:45:42
 * @LastEditTime: 2022-04-17 12:19:33
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\transforms\transformExpression.ts
 *
 */
import { NodeType } from "../ast";

// 处理多层包装
export function transformExpression(node) {
  if (node.type === NodeType.INTERPOLATION) {
    node.content = processExpression(node.content);
  }
}

function processExpression(node) {
  node.content = `_ctx.${node.content}`;

  return node;
}
