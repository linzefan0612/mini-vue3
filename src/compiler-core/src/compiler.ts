/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-20 20:47:43
 * @LastEditTime: 2022-04-20 20:49:25
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\src\compiler.ts
 *
 */

import { generate } from "./codegen";
import { baseParse } from "./parse";
import { transform } from "./transform";
import { transformElement } from "./transforms/transformElement";
import { transformExpression } from "./transforms/transformExpression";
import { transformText } from "./transforms/transformText";

// 暴露编译方法
export function baseCompile(template: string) {
  const ast = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  const code = generate(ast);
  return {
    code,
  };
}
