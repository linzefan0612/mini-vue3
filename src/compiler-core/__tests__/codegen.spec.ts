/*
 * @Author: Lin ZeFan
 * @Date: 2022-04-10 10:45:42
 * @LastEditTime: 2022-04-17 13:31:25
 * @LastEditors: Lin ZeFan
 * @Description:
 * @FilePath: \mini-vue3\src\compiler-core\__tests__\codegen.spec.ts
 *
 */
// import { generate } from "../src/codegen";

import { baseParse } from "../src/parse";
import { transform } from "../src/transform";
import { codegen } from "../src/codegen";
import { transformExpression } from "../src/transforms/transformExpression";
import { transformElement } from "../src/transforms/transformElement";
// import { transformText } from "../src/transforms/transformText";

test("text", () => {
  const template = "hi";
  const ast = baseParse(template);
  transform(ast);
  const code = codegen(ast);
  expect(code).toMatchSnapshot();
});

test("interpolation", () => {
  const template = "{{message}}";
  const ast = baseParse(template);
  transform(ast, {
    nodeTransforms: [transformExpression],
  });
  const code = codegen(ast);
  expect(code).toMatchSnapshot();
});

test("simple element", () => {
  const template = "<div></div>";
  const ast = baseParse(template);
  transform(ast, {
    // 加入处理插件
    nodeTransforms: [transformElement],
  });
  const code = codegen(ast);
  expect(code).toMatchSnapshot();
});

// test("element and interpolation", () => {
//   const ast = baseParse("<div>hi,{{msg}}</div>");
//   transform(ast, {
//     nodeTransforms: [transformElement, transformText, transformExpression],
//   });

//   const { code } = generate(ast);
//   expect(code).toMatchSnapshot();
// });
