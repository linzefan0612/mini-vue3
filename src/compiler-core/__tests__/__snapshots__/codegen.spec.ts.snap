// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpolation 1`] = `
"const { toDisplayString as _toDisplayString } = Vue
export function render(_ctx, _cache) { return _toDisplayString('_ctx.message') }"
`;

exports[`simple element 1`] = `
"const { createElementVNode as _createElementVNode } = Vue
export function render(_ctx, _cache) { return _createElementVNode(div, null,  }"
`;

exports[`text 1`] = `"export function render(_ctx, _cache) { return 'hi' }"`;

exports[`union 3 type 1`] = `
"const { createElementVNode as _createElementVNode, toDisplayString as _toDisplayString } = Vue
export function render(_ctx, _cache) { return _createElementVNode(div, null, 'hi,' + _toDisplayString('_ctx.message') }"
`;
